cpp/src/jit/parser.cpp

```diff
@@ -311,71 +280,41 @@ std::string ptx_parser::parse_param_list(const std::string& src)
 
 std::string ptx_parser::parse_function_header(const std::string& src)
 {
-  const size_t length = src.size();
-  size_t start        = 0;
-  size_t stop         = 0;
-
-  // Essentially we only need the information inside the two pairs of
-  // parentices.
-  while (start < length && is_white(src[start])) { start++; }
-
-  if (src[start] == '(') {  // This function has a return type
-    // First Pass: output param list
-    while (start < length && src[start] != '(') { start++; }
-    start++;
-    stop = start;
-    while (stop < length && src[stop] != ')') { stop++; }
-    stop++;
-  }
-
-  start = stop;
-  // The function name
-  while (start < length && is_white(src[start])) { start++; }
-  stop = start;
-  while (stop < length && !is_white(src[stop]) && src[stop] != '(') { stop++; }
-
-  start = stop;
-  // Second Pass: input param list
-  while (start < length && src[start] != '(') { start++; }
-  start++;
-  stop = start;
-  while (stop < length && src[stop] != ')') { stop++; }
-  std::string input_arg = parse_param_list(std::string(src, start, stop - start));
+  // Essentially we only need the information inside the two pairs of parentheses.
+  auto f = [&] {
+    auto i = std::find_if_not(src.cbegin(), src.cend(), [](auto c) { return is_white(c); });
+    if (*i == '(')  // This function has a return type
+      // First Pass: output param list
+      i = std::find_if_not(std::next(i), src.cend(), [](auto c) { return c == ')'; });
+    // The function name
+    i = std::find_if_not(std::next(i), src.cend(), [](auto c) { return is_white(c) || c == '('; });
+    // Second Pass: input param list
+    return std::next(std::find(i, src.cend(), '('));
+  }();
+
+  auto l = std::find(f, src.cend(), ')');
+
+  auto const input_arg = parse_param_list(std::string(f, l));
   return "\n__device__ __inline__ void " + function_name + "(" + input_arg + "){" + "\n";
 }
 
 std::string remove_comments(const std::string& src)
 {
-  // Remove the comments in the input ptx code.
-  size_t start       = 0;
-  size_t stop        = 0;
-  std::string output = src;
-  while (start < output.size() - 1) {
-    if (output[start] == '/' && output[start + 1] == '*') {
-      stop = start + 2;
-      while (stop < output.size() - 1) {
-        if (output[stop] == '*' && output[stop + 1] == '/') {
-          stop += 2;
-          break;
-        } else {
-          stop++;
-        }
-      }
-      output.erase(start, stop - start);
-    } else if (output[start] == '/' && output[start + 1] == '/') {
-      stop = start + 2;
-      while (stop < output.size()) {
-        if (output[stop] == '\n') {
-          // stop += 1; // Keep the newline here.
-          break;
-        } else {
-          stop++;
-        }
-      }
-      output.erase(start, stop - start);
-    } else {
-      start++;
-    }
+  std::string output;
```

Comment:
This looks like an improvement over the old code, but I think this could be better using iterators and algorithms as you did above.
Something like:
```cpp
auto f = src.cbegin();
while (f < src.cend()) {
  auto l = std::find(f, src.cend(), '/');
  output.append(f, l); // push chunk instead of 1 char at a time
  f = l;
  if (++l < src.cend()) {
    if (*l == '/') { // found "//"
      f = std::find(l, src.cend(), '\n'); // skip to end of line
    } else if (*l == '*') { // found "/*"
      // skip to end of next "*/"
      auto term = std::string("*/");
      f = std::search(std::next(l), src.cend, term.cbegin(), term.end()) + term.size();
    }
  }
}
```

Reply:
Oops, there was a bug in this suggestion because it didn't handle single `/` characters 😅 
```cpp
  auto f = src.cbegin();
  while (f < src.cend()) {
    auto l = std::find(f, src.cend(), '/');
    output.append(f, l); // push chunk instead of 1 char at a time
    f = std::next(l); // skip over '/'

    if (l < src.cend()) {
      char n = f < src.cend() ? *f : '\n';
      if (n == '/') { // found "//"
        f = std::find(f, src.cend(), '\n'); // skip to end of line
      } else if (n == '*') { // found "/*"
        // skip to end of next "*/"
        auto term = std::string("*/");
        f = std::search(std::next(f), src.cend(), term.cbegin(), term.cend()) + term.size();
      } else {
        // lone '/' should be pushed into output
        output.push_back('/');
      }
    }
  }
```

Reply:
Imo, i figured the algorithm we wanted just wasn't available. I want a `copy_if` with a custom stride length (what we want is three). Because I didn't have that, i figured raw loop is next best thing. I will sleep on your code. I am pretty sure I like yours better as well.

Reply:
Agree! Yours looks better. Updated :+1: